CRM/Core/Payment/PayPalImpl.php

```diff
@@ -702,15 +702,30 @@ static public function handlePaymentNotification() {
     if (is_numeric($lastParam)) {
       $params['processor_id'] = $lastParam;
     }
-    if (civicrm_api3('PaymentProcessor', 'getvalue', array(
-        'id' => $params['processor_id'],
-        'return' => 'class_name')
-      ) == 'Payment_PayPalImpl') {
-      $paypalIPN = new CRM_Core_Payment_PayPalIPN($params);
-    }
-    else {
-      $paypalIPN = new CRM_Core_Payment_PayPalProIPN($params);
+    $result = civicrm_api3('PaymentProcessor', 'get', array(
+      'sequential' => 1,
+      'id' => $params['processor_id'],
+      'api.PaymentProcessorType.getvalue' => array('return' => "name"),
+    ));
```

Comment:
As there should only be one processor returned when specifying an `id` why not use PaymentProcessor.getsingle, or drop the sequential and check that the ID exists in the returned `values` array.  Currently with this change, it will generate a PHP notice if `[values][0]` is undefined for some reason (eg. payment processor has been deleted).

Reply:
Fair point. What _should_ happen if the given payment processor has been deleted? I guess a fatal "not found" error is the best we could do, right?

Reply:
@twomice I'd suggest using `Civi::log()->error()` with a suitable error message.  Then return/exit or throw an exception rather than trigger a fatal error.  @eileenmcnaughton You'll probably have an opinion on this :-)

Reply:
I think it should fatal, either through an exception or a call to `fatal()`, so that PayPal can detect that the IPN is failing (and the `reporterror` extension can detect it and notify admins). In other words, I'm a bit uncomfortable with `Civi::log()->error()`, because it can easily go unnoticed.

Reply:
@mlutfy Could the reporterror extension not pick up Civi::log() style errors too?  I find it a real pain when you get lot's of fatal error info in the log files which is not always clear what the actual error is.  A single-line `Civi::log()->error()` can show exactly what the issue is if the message is written in a sensible way.

Reply:
For the extension, currently it doesn't, but you're right I should add support for it. However, I still think it should fatal so that PayPal can detect the error. It's not an error that should ever happen, and if it does, it should fail badly. It's not something where we're telling something polite to the user  and logging more details under the hood.

Reply:
@mlutfy Ok, makes sense

Reply:
Just on the error side of things - I think throwing an exception is generally better than 'fatal' because fatal should be seen as deprecated & Civi::Log-> is better than CRM_Core_Error::debug_log_message but I see the point that people aren't used to looking there & also Civi::Log should be pluggable in output. In general I'm not strongly on any side here

Reply:
I've replaced calls to CRM_Core_Error::debug_log_message() with exception; also added an exception if the processor can't be found by ID.